### PR TITLE
Preserve EA dtype in DataFrame.stack

### LINKED ISSUE #23077: Handle ExtensionArrays in Series.unstack / DataFrame.stack
Here's a test for `Series.unstack`

```diff
diff --git a/pandas/tests/extension/base/reshaping.py b/pandas/tests/extension/base/reshaping.py
index 0340289e0..a040fba63 100644
--- a/pandas/tests/extension/base/reshaping.py
+++ b/pandas/tests/extension/base/reshaping.py
@@ -171,3 +171,16 @@ class BaseReshapingTests(BaseExtensionTests):
                  [data[0], data[0], data[1], data[2], na_value],
                  dtype=data.dtype)})
         self.assert_frame_equal(res, exp[['ext', 'int1', 'key', 'int2']])
+
+    def test_unstack(self, data):
+        data = data[:4]
+        ser = pd.Series(
+            data,
+            index=pd.MultiIndex.from_product([["A", "B"],
+                                              ["a", "b"]]),
+        )
+        result = ser.unstack()
+        expected = pd.DataFrame({"a": data.take([0, 2]),
+                                 "b": data.take([1, 3])},
+                                index=['A', 'B'])
+        self.assert_frame_equal(result, expected)
```

We don't do so well right now. Just categorical passes. 
```
========================================================================= FAILURES =========================================================================
/Users/taugspurger/sandbox/pandas/pandas/core/reshape/reshape.py:112: IndexError: tuple index out of range
/Users/taugspurger/sandbox/pandas/pandas/core/reshape/reshape.py:112: IndexError: tuple index out of range
/Users/taugspurger/sandbox/pandas/pandas/core/reshape/reshape.py:112: IndexError: tuple index out of range
/Users/taugspurger/sandbox/pandas/pandas/core/reshape/reshape.py:112: IndexError: tuple index out of range
/Users/taugspurger/sandbox/pandas/pandas/core/reshape/reshape.py:112: IndexError: tuple index out of range
/Users/taugspurger/sandbox/pandas/pandas/core/reshape/reshape.py:112: IndexError: tuple index out of range
/Users/taugspurger/sandbox/pandas/pandas/core/reshape/reshape.py:112: IndexError: tuple index out of range
/Users/taugspurger/sandbox/pandas/pandas/core/reshape/reshape.py:112: IndexError: tuple index out of range
/Users/taugspurger/sandbox/pandas/pandas/core/reshape/reshape.py:204: AttributeError: 'IntervalArray' object has no attribute 'reshape'
/Users/taugspurger/sandbox/pandas/pandas/tests/extension/decimal/array.py:55: TypeError: All values must be of type <class 'decimal.Decimal'>
/Users/taugspurger/sandbox/pandas/pandas/tests/extension/json/array.py:88: TypeError: list indices must be integers or slices, not NoneType
```

No test for DataFrame.stack. In https://github.com/pandas-dev/pandas/pull/22862/files there's a WIP for stack that's based around `ExtensionArray._concat_same_type` instead of `.reshape`

### PR DESCRIPTION
**Pull request by TomAugspurger**
closes #23077

There were two bugs in master (not present in 0.23.4), probably from the SparseArray PR

1. We need to unbox the EA values from Series before passing to `EA._concat_same_type`
2. We need to followup with a `take` to get the correct order.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Just a WIP for now. I feel like we're lacking on tests here. Will add one for stacking a single level from a dataframe with a MultiIndex in the columns.

File: pandas/core/reshape/reshape.py [hunk 1]
```diff
@@ -470,8 +470,15 @@ def factorize(index):
...
             arr = dtype.construct_array_type()
             new_values = arr._concat_same_type([
-                col for _, col in frame.iteritems()
+                col._values for _, col in frame.iteritems()
             ])
+            # final take to get the order correct.
+            # idx is an indexer like
+            # [c0r0, c1r0, c2r0, ...,
+            #  c0r1, c1r1, c241, ...]
+            idx = np.arange(N * K).reshape(K, N).T.ravel()
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: why is this needed here?

Comment (TomAugspurger) [reply pandas/core/reshape/reshape.py]: Had the order backwards on master :/ Without it, we were returning

```
In [6]: df.stack()
Out[6]:
0  A    a
   B    b
1  A    a
   B    b
dtype: category
Categories (2, object): [a, b]
```

instead of the correct

```
In [6]: df.stack()
Out[6]:
0  A    a
   B    a
1  A    b
   B    b
dtype: category
Categories (2, object): [a, b]
```

**Comment by TomAugspurger** [comment]
Just fixed the decimal failures.

There will be a remaining test failure I haven't addressed yet. We had a test that did `Series[categorical].unstack(fill_value=value)` for a value that wasn't part of the original `categories` in the categorical. We aren't `take` correctly right now (https://github.com/pandas-dev/pandas/issues/23296), but once that's fixed there's still an API discussion: should we allow `take` to take "new" categories that weren't previously present.

```python
In [2]: cat = pd.Categorical(['a', 'a', 'b'])

In [3]: cat.take([0, -1, -1], fill_value='d', allow_fill=True)
```

Should that raise? Return a Categorical with categories `['a', 'b', 'd']`?

I'm having some deja vu right now; I think we've discussed this before.

I think if we were designing that today, we wouldn't have allowed that. Do we want to put

**Comment by TomAugspurger** [comment]
"Fixed" the sparse values. We were failing to handle `DataFrame[SparseArray].astype(object)` correctly. On master, Series[sparse].astype(object) / Frame[sparse].astype(object) is sparse, but I think we want to change that. https://github.com/pandas-dev/pandas/issues/23125

I'll probably do that today, and then pick up the reshaping PRs afterwards.

**Review by jorisvandenbossche — COMMENTED** [review]
Is this still WIP ? 
It looks good to me.

**Comment by TomAugspurger** [comment]
I think we need tests where the columns are a MultiIndex.

________________________________

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -633,7 +634,9 @@ def _astype(self, dtype, copy=False, errors='raise', values=None,
             return self
 
         if klass is None:
-            if dtype == np.object_:
+            # sparse is "special" and preserves sparsity.
+            # We're changing this in GH-23125
+            if dtype == np.object_ and is_sparse(values):
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: use is_object_dtype

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py [hunk 3]
```diff
@@ -632,7 +633,10 @@ def _astype(self, dtype, copy=False, errors='raise', values=None,
             return self
 
         if klass is None:
-            if dtype == np.object_:
+            if is_sparse(self.values):
+                # Series[Sparse].astype(object) is sparse.
+                klass = ExtensionBlock
+            elif is_object_dtype(dtype):
                 klass = ObjectBlock
             elif is_extension_array_dtype(dtype):
```
Comment (jreback) [inline pandas/core/internals/blocks.py]: so maybe should just move the ``is_extension_array_dtype`` up to first, and add a ``is_extension_dtype(self.values)`` test as well (should encompas your is_sparse check) and is more general

File: pandas/core/reshape/reshape.py [hunk 4]
```diff
@@ -604,12 +611,28 @@ def _convert_level_number(level_num, columns):
             chunk.columns = level_vals.take(chunk.columns.labels[-1])
             value_slice = chunk.reindex(columns=level_vals_used).values
         else:
-            if frame._is_mixed_type:
+            if (frame._is_homogeneous_type and
+                    is_extension_array_dtype(frame.dtypes.iloc[0])):
+                dtype = this.loc[:, this.columns[loc]].dtypes.iloc[0]
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: ``this[this.columns[loc]]`` does this not work?

File: pandas/core/reshape/reshape.py [hunk 5]
```diff
@@ -604,12 +611,28 @@ def _convert_level_number(level_num, columns):
...
+                subset = this.loc[:, this.columns[loc]]
+
+                value_slice = dtype.construct_array_type()._concat_same_type(
+                    [x._values for _, x in subset.iteritems()]
+                )
+                N, K = this.shape
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: maybe make a little helper function out of these 3 lines (so you can use it above as well)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/blocks.py (hunk 3 above)
Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: I'll make that change and run the test suite.

I was kinda worried about "false positives" here, but I suppose it's exactly what we want if an extension array claims it's object dtype.

Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: As posted in the unstack PR, we need to special case Space here, since it's the only (internal) extension type that has special `.astype(object)` behavior.

File: doc/source/whatsnew/v0.24.0.txt [hunk 6]
```diff
@@ -849,7 +849,7 @@ update the ``ExtensionDtype._metadata`` tuple to match the signature of your
 - Updated the ``.type`` attribute for ``PeriodDtype``, ``DatetimeTZDtype``, and ``IntervalDtype`` to be instances of the dtype (``Period``, ``Timestamp``, and ``Interval`` respectively) (:issue:`22938`)
 - :func:`ExtensionArray.isna` is allowed to return an ``ExtensionArray`` (:issue:`22325`).
 - Support for reduction operations such as ``sum``, ``mean`` via opt-in base class method override (:issue:`22762`)
-- :meth:`Series.unstack` no longer converts extension arrays to object-dtype ndarrays. The output ``DataFrame`` will now have the same dtype as the input. This changes behavior for Categorical and Sparse data (:issue:`23077`).
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.txt]: This was accidentally added in the PeriodArray PR. Will be implemented for good in https://github.com/pandas-dev/pandas/pull/23284

File: pandas/tests/extension/base/reshaping.py [hunk 7]
```diff
@@ -170,3 +170,25 @@ def test_merge(self, data, na_value):
...
+                                  names=['outer', 'inner']),
+    ])
+    def test_stack(self, data, columns):
+        df = pd.DataFrame({"A": data[:5], "B": data[:5]})
+        df.columns = columns
+        result = df.stack()
+        expected = df.astype(object).stack()
+        # we need a second astype(object), in case the constructor inferred
+        # object -> specialized, as is done for period.
+        expected = expected.astype(object)
```
Comment (TomAugspurger) [inline pandas/tests/extension/base/reshaping.py]: This is kinda strange. For `DataFrame[ndarray[object]].stack()` of all periods, we actually infer `period`-dtype. Do we want that, or should we explicitly pass `dtype=object` when creating the new series / frame to ensure that we don't infer the "correct" dtype?

```python
In [1]: import pandas as pd

In [2]: a = pd.core.arrays.period_array(['2000', '2001'], freq='D')

In [3]: pd.DataFrame({"A": a, "B": a}).astype(object).dtypes
Out[3]:
A    object
B    object
dtype: object

In [4]: pd.DataFrame({"A": a, "B": a}).astype(object).stack().dtype
Out[4]: period[D]
```

(that's on master)

**Comment by TomAugspurger** [comment]
All green.

**Comment by TomAugspurger** [comment]
Merged master to fix the merge conflict with the unstack PR.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm.
